src/main/java/uk/ac/ic/wlgitbridge/bridge/lock/InProcessLock.java

```diff
@@ -0,0 +1,85 @@
+package uk.ac.ic.wlgitbridge.bridge.lock;
+
+import uk.ac.ic.wlgitbridge.bridge.util.Pair;
+import uk.ac.ic.wlgitbridge.util.Log;
+
+import java.util.HashMap;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * A thin wrapper around a ReentrantLock
+ */
+public class InProcessLock implements ProjectLock {
+
+  // Static map of underlying locks, for reference counting
+  private static HashMap<String, Pair<ReentrantLock, Integer>> countedLocks =
+          new HashMap<>();
+
+  private String projectName;
+  private ReentrantLock lock;
+  private OperationTracker registry;
+
+  public InProcessLock(String projectName, OperationTracker operationTracker) {
+    this.projectName = projectName;
+    this.registry = operationTracker;
+    this.lock = getBaseLock(projectName);
+  }
+
+  private synchronized ReentrantLock getBaseLock(String projectName) {
+    Pair<ReentrantLock, Integer> lockPair;
+    if (InProcessLock.countedLocks.containsKey(projectName)) {
+      lockPair = InProcessLock.countedLocks.get(projectName);
+      Integer newCount = lockPair.updateRight((Integer i) -> ++i); // DEBUG
```

Comment:
`// DEBUG`?

Reply:
A debug statement that grew up to become a real statement :) I'll remove the comment :+1: